packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -6445,6 +6477,71 @@ class _NullWidget extends Widget {
   Element createElement() => throw UnimplementedError();
 }
 
+/// A class that is used to record how widgets are built on each frame.
+class BuildRecorder {
+  final List<_RecorderFrameData> _frameData = <_RecorderFrameData>[];
+  _RecorderFrameData _currentFrame = _RecorderFrameData();
+
+  void _recordElement(Element element) {
+    final Widget? parent = element._parent?.widget;
+    final Widget widget = element.widget;
+    if (parent == null) {
+       _currentFrame.widgets[widget] = null;
+       return;
+    }
+    _currentFrame.widgets[widget] = parent;
+  }
+
+  void _recordRenderObjectElement(RenderObjectElement element) {
+    final Widget? parent = element._parent?.widget;
+    final Widget widget = element.widget;
+    if (parent == null) {
+       _currentFrame.widgets[widget] = null;
+       return;
+    }
+    _currentFrame.renderObjects[widget] = element.renderObject;
+    _currentFrame.widgets[widget] = parent;
+  }
+
+  /// Finish the recording of the current frame and pepare the next frame object.
+  void finishFrame(int frameNumber) {
+    _currentFrame.id = frameNumber;
+    _frameData.add(_currentFrame);
+    _currentFrame = _RecorderFrameData();
+  }
+
+  /// Convert the recorded frame data into a JSON object.
+  Map<String, Object> toJson() {
+    return <String, Object>{
+      'frames': <Object>[
+        for (_RecorderFrameData data in _frameData)
+          data.toJson(),
+      ],
+    };
+  }
+}
+
+class _RecorderFrameData {
+  _RecorderFrameData();
+
+  int id = -1;
+  final Map<Widget, Widget?> widgets = HashMap<Widget, Widget?>();
+  final Map<Widget, RenderObject> renderObjects = HashMap<Widget, RenderObject>();
+
+  Map<String, Object> toJson() {
+    return <String, Object>{
+      'id': id,
+      'widgets': <String, String?>{
+        for (MapEntry<Widget, Widget?> entry in widgets.entries)
+          '${entry.key.runtimeType}|${entry.key.hashCode}': entry.value == null
```

Comment:
side note: If a widget rebuilds the next frame it will likely have a new hashCode since widget objects are short lived. So this may not be the most useful way of identifying a widget over multiple frames...

Also, for consistency could you use https://master-api.flutter.dev/flutter/foundation/describeIdentity.html here?